staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/validator.go

```diff
@@ -18,298 +18,83 @@ package validatingadmissionpolicy
 }
 
-func objectToResolveVal(r runtime.Object) (interface{}, error) {
-	if r == nil || reflect.ValueOf(r).IsNil() {
-		return nil, nil
-	}
-	v, err := convertObjectToUnstructured(r)
-	if err != nil {
-		return nil, err
-	}
-	return v.Object, nil
-}
-
-func policyDecisionActionForError(f v1alpha1.FailurePolicyType) PolicyDecisionAction {
-	if f == v1alpha1.Ignore {
+func policyDecisionActionForError(f v1.FailurePolicyType) PolicyDecisionAction {
+	if f == v1.Ignore {
 		return ActionAdmit
 	}
 	return ActionDeny
 }
 
-// Validate validates all cel expressions in Validator and returns a PolicyDecision for each CEL expression or returns an error.
-// An error will be returned if failed to convert the object/oldObject/params/request to unstructured.
-// Each PolicyDecision will have a decision and a message.
-// policyDecision.message will be empty if the decision is allowed and no error met.
-func (v *CELValidator) Validate(versionedAttr *generic.VersionedAttributes, versionedParams runtime.Object) ([]PolicyDecision, error) {
-	// TODO: replace unstructured with ref.Val for CEL variables when native type support is available
-	decisions := make([]PolicyDecision, len(v.compilationResults))
-	var err error
-
-	oldObjectVal, err := objectToResolveVal(versionedAttr.VersionedOldObject)
-	if err != nil {
-		return nil, err
-	}
-	objectVal, err := objectToResolveVal(versionedAttr.VersionedObject)
-	if err != nil {
-		return nil, err
-	}
-	paramsVal, err := objectToResolveVal(versionedParams)
-	if err != nil {
-		return nil, err
+// Validate takes a list of Evaluation and a failure policy and converts them into actionable PolicyDecisions
+func (v *validator) Validate(versionedAttr *generic.VersionedAttributes, versionedParams runtime.Object) []PolicyDecision {
+	var f v1.FailurePolicyType
+	if v.failPolicy == nil {
+		f = v1.Fail
+	} else {
+		f = *v.failPolicy
 	}
 
-	request := createAdmissionRequest(versionedAttr.Attributes)
-	requestVal, err := convertObjectToUnstructured(request)
+	evalResults, err := v.filter.ForInput(versionedAttr, versionedParams, cel.CreateAdmissionRequest(versionedAttr.Attributes))
 	if err != nil {
-		return nil, err
-	}
-	va := &validationActivation{
-		object:    objectVal,
-		oldObject: oldObjectVal,
-		params:    paramsVal,
-		request:   requestVal.Object,
-	}
-
-	var f v1alpha1.FailurePolicyType
-	if v.policy.Spec.FailurePolicy == nil {
-		f = v1alpha1.Fail
-	} else {
-		f = *v.policy.Spec.FailurePolicy
+		return []PolicyDecision{
+			{
+				Action:     policyDecisionActionForError(f),
+				Evaluation: EvalError,
+				Message:    err.Error(),
+			},
+		}
 	}
+	decisions := make([]PolicyDecision, len(evalResults))
 
-	for i, compilationResult := range v.compilationResults {
-		validation := v.policy.Spec.Validations[i]
-
-		var policyDecision = &decisions[i]
+	for i, evalResult := range evalResults {
+		var decision = &decisions[i]
+		var validation = evalResult.ExpressionAccessor.(*ValidationCondition)
```

Comment:
Note that this can panic if the type cast fails.

Because the code is written such that you get back what you put in, the risk is currently low.

Using generics or the `newType, ok := var.(typecast); if !ok {...}` model would prevent a panic (in this case probably raise an error).

Simpler to do the later, but we should consider generics as a follow-on

Reply:
I'll convert to generics on the next run :)